core/kernel/persistence/starsql/class.StarRdf.php

```diff
@@ -60,97 +43,117 @@ protected function getPersistence()
+     * {@inheritDoc}
      */
-    public function search($predicate, $object)
-    {
-        throw new common_Exception('Not implemented');
-    }
-
     public function getIterator()
     {
-        throw new common_Exception('Not implemented! ' . __FILE__ . ' line: ' . __LINE__);
+        return new RecursiveIteratorIterator(new core_kernel_persistence_starsql_StarIterator($this->model));
     }
 
     /**
-     * @return Ontology
+     * @param iterable $tripleList
+     * @param Format $format
+     *
+     * @return string
      */
-    protected function getModel()
+    private function triplesToValues(iterable $tripleList, Format $format): string
     {
-        return $this->model;
-    }
+        $graph = new Graph();
+
+        /** @var core_kernel_classes_Triple $triple */
+        foreach ($tripleList as $triple) {
+            if (!empty($triple->lg)) {
+                $graph->addLiteral(
+                    $triple->subject,
+                    $triple->predicate,
+                    $triple->object,
+                    $triple->lg
+                );
+            } elseif (\common_Utils::isUri($triple->object)) {
+                $graph->addResource($triple->subject, $triple->predicate, $triple->object);
+            } else {
+                $graph->addLiteral($triple->subject, $triple->predicate, $triple->object);
+            }
+        }
 
-    /**
-     * @param core_kernel_classes_Triple $triple
-     * @return array
-     */
-    protected function tripleToValue(core_kernel_classes_Triple $triple): array
-    {
-        return [
-            'modelid' => $triple->modelid,
-            'subject' => $triple->subject,
-            'predicate' => $triple->predicate,
-            'object' => $triple->object,
-            'l_language' => is_null($triple->lg) ? '' : $triple->lg,
-            'author' => is_null($triple->author) ? '' : $triple->author,
-            'epoch' => $this->getPersistence()->getPlatForm()->getNowExpression()
-        ];
+        return $graph->serialise($format);
     }
 
-    protected function getTripleParameterTypes(): array
+    private function createSystemTripleQuery(iterable $tripleList): ?Statement
     {
-        return self::TRIPLE_PARAMETER_TYPE;
+        $systemSubjectList = [];
+        /** @var core_kernel_classes_Triple $triple */
+        foreach ($tripleList as $triple) {
+            if (!empty($triple->modelid) && $triple->modelid != \core_kernel_persistence_starsql_StarModel::DEFAULT_WRITABLE_MODEL) {
+                $systemSubjectList[$triple->subject] = true;
+            }
+        }
+
+        $query  = null;
+        if (!empty($systemSubjectList)) {
+            $systemNode = Query::node('Resource');
+            $query = Query::new()->match($systemNode)
+                ->where($systemNode->property('uri')->in(array_keys($systemSubjectList)))
+                ->set($systemNode->labeled('System'));
+        }
+
+        return Statement::create($query->build());
```

Comment:
If `$systemSubjectList` is empty, then `$query` remains to be `null`, and this line errors out

Reply:
thank you good catch